### PR TITLE
Create new Remove server CLI command

### LINKED ISSUE #1304: Implement remove server functionality for IML CLI
Implement the ability to remove servers via the IML CLI.

We can currently add new servers using a hostlist expression. For example: 

```sh
iml server add --hosts mds[1,2].local,oss[1,2].local --profile base_managed_patchless
```
will deploy the IML agent and set a profile for mds1.local, mds2.local, oss1.local, oss2.local in parallel.

What we are currently missing is the equivalent removal syntax. For example:

```sh
iml server remove --hosts mds[1,2].local,oss[1,2].local
```

should remove the agent from mds1.local, mds2.local, oss1.local, oss2.local in parallel.

The code for the server CLI lives here: 

https://github.com/whamcloud/integrated-manager-for-lustre/blob/25587229ee99e419fd8083469ab78436641c773c/iml-manager-cli/src/server.rs

To implement, a new Enum variant `ServerCommand::Remove(config)` should be created. When that variant is matched, it will:

1. Parse the hostlist out of the config and return on error
1. Fetch the current known hosts from the API and filter out any known ones (printing message on stdout that they have been filtered).
1. Ensure the hosts can actually be removed by checking the /action endpoint and providing the relevant composite_ids. Each server that can be removed should have a `Remove` verb.
1. Remove the hosts by issuing a state change job to each host serially. This is a PUT to the resource_uri of each host resource to remove.

### PR DESCRIPTION
**Pull request by no111u3**
Create a remove server cline that can remove n server in parallel using hostlist expressions.

This first revision is limited in scope:

Only existing keys are supported.
Any issue in removing will abort the command.
Fixes #1304

---
This change is [Reviewable](https://reviewable.io/reviews/whamcloud/integrated-manager-for-lustre/1437)

### CODE REVIEW

File: iml-manager-cli/src/server.rs [hunk 1]
```diff
@@ -23,6 +23,13 @@ use std::{
 use structopt::StructOpt;
 use tokio::time::delay_for;
 
+#[derive(StructOpt, Debug)]
+pub struct RemoveHosts {
+    /// The host(s) to update. Takes a hostlist expression
+    #[structopt(short = "h", long = "hosts")]
```
Comment (ip1981) [inline iml-manager-cli/src/server.rs]: Might conflict with `-h` for help.

File: iml-manager-cli/src/server.rs [hunk 2]
```diff
@@ -41,6 +48,9 @@ pub enum ServerCommand {
     /// Add new servers to IML
     #[structopt(name = "add")]
     Add(AddHosts),
+    /// Remove setted servers from IML
```
Comment (ip1981) [inline iml-manager-cli/src/server.rs]:
```suggestion
    /// Remove servers from IML
```

File: iml-manager-cli/src/server.rs [hunk 3]
```diff
@@ -373,6 +391,135 @@ pub async fn server_cli(command: ServerCommand) -> Result<(), ImlManagerCliError
...
+                                    format_error("Fail")
+                                };
+
+                                vec![x.name.clone(), v]
+                            }),
+                        );
+
+                        false
+                    } else {
+                        state && true
```
Comment (ip1981) [inline iml-manager-cli/src/server.rs]: `&& true` is no-op.

File: iml-manager-cli/src/server.rs [hunk 4]
```diff
@@ -373,6 +391,135 @@ pub async fn server_cli(command: ServerCommand) -> Result<(), ImlManagerCliError
...
+                .json()
+                .await
+                .map_err(iml_manager_client::ImlManagerClientError::Reqwest)?;
+
+            let cmds = objects
+                .into_iter()
+                .map(|CmdWrapper { command }| command)
+                .collect();
+
+            term.write_line(&format!("{} preflight checks...", style("Running").green()))?;
```
Comment (ip1981) [inline iml-manager-cli/src/server.rs]: Are any checks really needed before removing the servers?

File: iml-manager-cli/src/server.rs (hunk 2 above)
Comment (no111u3) [reply iml-manager-cli/src/server.rs]: Thanks, I fix it later.

File: iml-manager-cli/src/server.rs (hunk 4 above)
Comment (no111u3) [reply iml-manager-cli/src/server.rs]: IMO not needed, but in original issue it marked as needed.

File: iml-manager-cli/src/server.rs (hunk 3 above)
Comment (no111u3) [reply iml-manager-cli/src/server.rs]: Removed.

File: iml-manager-cli/src/server.rs (hunk 1 above)
Comment (no111u3) [reply iml-manager-cli/src/server.rs]: Changed to `-d` and `deleted_hosts`

Comment (ip1981) [reply iml-manager-cli/src/server.rs]: Maybe use positional arguments instead of options: `... remove <host expression>`?

Comment (no111u3) [reply iml-manager-cli/src/server.rs]: I only reproduce code of @jgrund, so I think it may be generally and better than now. I can create pr for change adding hosts too.

File: iml-manager-cli/src/server.rs [hunk 5]
```diff
@@ -373,6 +391,135 @@ pub async fn server_cli(command: ServerCommand) -> Result<(), ImlManagerCliError
...
+                return Err(ImlManagerCliError::ApiError(
+                    "Preflight checks failed".into(),
+                ));
+            }
+
+            let host_ids: Vec<_> = cmds
+                .into_iter()
+                .flat_map(|cmd| cmd.jobs)
+                .filter_map(|job| extract_api_id(&job).map(|x| x.to_string()))
+                .collect();
```
Comment (mkpankov) [inline iml-manager-cli/src/server.rs]: Intermediate `.collect()` isn't needed here, since the iterator is directly used in next step for `.map()`.

File: iml-manager-cli/src/main.rs [hunk 6]
```diff
@@ -49,12 +49,12 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
 
     tracing::subscriber::set_global_default(subscriber).unwrap();
 
-    dotenv::from_path("/var/lib/chroma/iml-settings.conf").expect("Could not load cli env");
-
     let matches = App::from_args();
 
     tracing::debug!("Matching args {:?}", matches);
 
+    dotenv::from_path("/var/lib/chroma/iml-settings.conf").expect("Could not load cli env");
```
Comment (mkpankov) [inline iml-manager-cli/src/main.rs]: Why this change?

Comment (no111u3) [reply iml-manager-cli/src/main.rs]: Because application doesn't generate panic for `-h` request

File: iml-manager-cli/src/server.rs (hunk 5 above)
Comment (no111u3) [reply iml-manager-cli/src/server.rs]: But objects needs to store before iteration.

Comment (no111u3) [reply iml-manager-cli/src/server.rs]: Fixed.

File: iml-manager-cli/src/server.rs [hunk 7]
```diff
@@ -373,6 +391,133 @@ pub async fn server_cli(command: ServerCommand) -> Result<(), ImlManagerCliError
...
             wait_for_cmds(cmds).await?;
         }
+        ServerCommand::Remove(config) => {
+            let term = Term::stdout();
+
+            let remove_hosts = hostlist_parser::parse(&config.hosts)?;
+
+            tracing::debug!("Parsed hosts {:?}", remove_hosts);
+
+            let (_, api_hosts) = future::try_join(get_hosts(), get_hosts()).await?;
```
Comment (jgrund) [inline iml-manager-cli/src/server.rs]: You are making two API calls here and throwing one away...

```suggestion
            let api_hosts = get_hosts().await?;
```

File: iml-manager-cli/src/server.rs [hunk 8]
```diff
@@ -373,6 +391,133 @@ pub async fn server_cli(command: ServerCommand) -> Result<(), ImlManagerCliError
...
+
+            let known_hosts = filter_known_hosts(&remove_hosts, &api_hosts.objects);
```
Comment (jgrund) [inline iml-manager-cli/src/server.rs]: This is the opposite of what we want.

We want to filter unknown hosts. This will filter known hosts.

Comment (jgrund) [reply iml-manager-cli/src/server.rs]: We also need to print a message about any hosts that were supposed to be removed, but did not exist

File: iml-manager-cli/src/server.rs [hunk 9]
```diff
@@ -373,6 +391,133 @@ pub async fn server_cli(command: ServerCommand) -> Result<(), ImlManagerCliError
...
+            if !all_passed {
+                return Err(ImlManagerCliError::ApiError(
+                    "Preflight checks failed".into(),
+                ));
+            }
+
+            let host_ids = cmds
+                .into_iter()
+                .flat_map(|cmd| cmd.jobs)
+                .filter_map(|job| extract_api_id(&job).map(|x| x.to_string()));
```
Comment (jgrund) [inline iml-manager-cli/src/server.rs]: None of this is needed to remove hosts.

File: iml-manager-cli/src/server.rs [hunk 10]
```diff
@@ -373,6 +391,133 @@ pub async fn server_cli(command: ServerCommand) -> Result<(), ImlManagerCliError
...
+
+            let objects = host_ids
+                .map(|address| HostRemoverConfig {
+                    address,
+                    verb: "Remove".to_string(),
+                    auth_type: "existing_keys_choice".into(),
```
Comment (jgrund) [inline iml-manager-cli/src/server.rs]: Have you tested whether `auth_type` and `server_profile` are needed?

I haven't checked, but I'd be surprised if they were required to remove a host...

File: iml-manager-cli/src/server.rs [hunk 11]
```diff
@@ -23,6 +23,13 @@ use std::{
 use structopt::StructOpt;
 use tokio::time::delay_for;
 
+#[derive(StructOpt, Debug)]
+pub struct RemoveHosts {
+    /// The host(s) to update. Takes a hostlist expression
```
Comment (jgrund) [inline iml-manager-cli/src/server.rs]:
```suggestion
    /// The host(s) to remove. Takes a hostlist expression
```

File: iml-manager-cli/src/server.rs (hunk 9 above)
Comment (jgrund) [reply iml-manager-cli/src/server.rs]: https://github.com/whamcloud/integrated-manager-for-lustre/blob/644624fffa92e35ac16a82951a17c8bcbc78e2b5/iml-manager-cli/src/server.rs#L405-L480

File: iml-manager-cli/src/server.rs (hunk 7 above)
Comment (no111u3) [reply iml-manager-cli/src/server.rs]: Done.

File: iml-manager-cli/src/server.rs (hunk 8 above)
Comment (no111u3) [reply iml-manager-cli/src/server.rs]: Done

File: iml-manager-cli/src/server.rs (hunk 9 above)
Comment (no111u3) [reply iml-manager-cli/src/server.rs]: Done.

File: iml-manager-cli/src/server.rs (hunk 11 above)
Comment (no111u3) [reply iml-manager-cli/src/server.rs]: Done.

File: iml-manager-cli/src/server.rs [hunk 12]
```diff
@@ -373,6 +399,87 @@ pub async fn server_cli(command: ServerCommand) -> Result<(), ImlManagerCliError
...
+            let remove_hosts = hostlist_parser::parse(&config.hosts)?;
+
+            tracing::debug!("Parsed hosts {:?}", remove_hosts);
+
+            let api_hosts = get_hosts().await?;
+
+            let unknown_hosts = filter_unknown_hosts(&remove_hosts, &api_hosts.objects);
+
+            for unknown_host in unknown_hosts {
+                display_cancelled(format!("Host {} unknown in IML. Please remove first if attempting to complete removement.", unknown_host.fqdn));
```
Comment (jgrund) [inline iml-manager-cli/src/server.rs]:
```suggestion
                display_cancelled(format!("Host {} unknown to IML and will not be removed.", unknown_host.fqdn));
```

File: iml-manager-cli/src/server.rs (hunk 9 above)
Comment (jgrund) [reply iml-manager-cli/src/server.rs]: Some unused code is still there (lines shown in previous comment).

File: iml-manager-cli/src/server.rs (hunk 12 above)
Comment (no111u3) [reply iml-manager-cli/src/server.rs]: Done.

File: iml-manager-cli/src/server.rs [hunk 13]
```diff
@@ -419,6 +445,86 @@ pub async fn server_cli(command: ServerCommand) -> Result<(), ImlManagerCliError
...
+            let cmds = objects
+                .into_iter()
+                .map(|CmdWrapper { command }| command)
+                .collect();
+
+            term.write_line(&format!("{} preflight checks...", style("Running").green()))?;
+
+            let _cmds = wait_for_cmds(cmds).await?;
+
+            let objects = remove_hosts
```
Comment (ip1981) [inline iml-manager-cli/src/server.rs]: It looks like all the hosts, even the unknown ones, will be queried. Will the request be successful?

File: iml-manager-cli/src/server.rs [hunk 14]
```diff
@@ -419,6 +445,86 @@ pub async fn server_cli(command: ServerCommand) -> Result<(), ImlManagerCliError
...
+                    .json()
+                    .await
+                    .map_err(iml_manager_client::ImlManagerClientError::Reqwest)?;
+
+            tracing::debug!("command and hosts {:?}", objects);
+
+            let (commands, hosts): (Vec<_>, Vec<_>) = objects
+                .into_iter()
+                .filter_map(|x| x.command_and_host)
+                .map(|x| (x.command, x.host))
```
Comment (ip1981) [inline iml-manager-cli/src/server.rs]: I am not sure, but probably this `map` and `the above `filter_map` can be merge in one `filter_map`.

File: iml-manager-cli/src/server.rs [hunk 15]
```diff
@@ -419,6 +445,86 @@ pub async fn server_cli(command: ServerCommand) -> Result<(), ImlManagerCliError
...
+            let host_ids: Vec<_> = hosts
+                .iter()
+                .map(|x| ["id__in".into(), x.id.to_string()])
+                .chain(iter::once(["limit".into(), "0".into()]))
+                .collect();
+
+            let ApiList { objects, .. }: ApiList<HostProfileWrapper> =
+                get(HostProfile::endpoint_name(), host_ids).await?;
+
+            tracing::debug!("Host Profiles {:?}", objects);
```
Comment (ip1981) [inline iml-manager-cli/src/server.rs]: It looks like the above API call is here only to show this debug message?

File: iml-manager-cli/src/server.rs [hunk 16]
```diff
@@ -419,6 +445,86 @@ pub async fn server_cli(command: ServerCommand) -> Result<(), ImlManagerCliError
...
+            wait_for_cmds(commands).await?;
+
+            let host_ids: Vec<_> = hosts
+                .iter()
+                .map(|x| ["id__in".into(), x.id.to_string()])
+                .chain(iter::once(["limit".into(), "0".into()]))
+                .collect();
+
+            let ApiList { objects, .. }: ApiList<HostProfileWrapper> =
+                get(HostProfile::endpoint_name(), host_ids).await?;
```
Comment (ip1981) [inline iml-manager-cli/src/server.rs]: I may misunderstand, but if the hosts are removed, will this call make any sense?

File: iml-manager-cli/src/server.rs (hunk 13 above)
Comment (no111u3) [reply iml-manager-cli/src/server.rs]: Done.

File: iml-manager-cli/src/server.rs (hunk 14 above)
Comment (no111u3) [reply iml-manager-cli/src/server.rs]: It will be not pretty for reading.

File: iml-manager-cli/src/server.rs (hunk 15 above)
Comment (no111u3) [reply iml-manager-cli/src/server.rs]: I copied it from Add operation

File: iml-manager-cli/src/server.rs (hunk 16 above)
Comment (no111u3) [reply iml-manager-cli/src/server.rs]: I misunderstand it too, but It added by point 4 in #1304.

File: iml-manager-cli/src/server.rs (hunk 15 above)
Comment (no111u3) [reply iml-manager-cli/src/server.rs]: Yes, sure.

**Comment by jgrund** [comment]
Thanks for submitting this @no111u3, I have pushed a few updates to this patch after testing. 

remove-servers
